### ISSUE TITLE
Automatic backup via WebDAVS

### ISSUE THREAD

**Issue by Addy90** [opnsense/core#2250]
@fabianfrz helped me figuring out in [plugins#591](https://github.com/opnsense/plugins/issues/591) what is the case:

What I want to suggest as a feature is an automatic backup sync not only to proprietary Google Drive but also for WebDAVS backends like Nextcloud / ownCloud (as I suppose, many people self-hosting Open Source Firewalls might self-host Open Source "Cloud-Drives", too, like I do.)

Alternatives for WebDAVS would have been FTP / SMB / NFS so that the WebDAVS Desktop Client from Nextcloud / ownCloud can do the rest if WebDAVS is too complex (for example with a local FTP installed on a Windows machine, like I do when Network Printers at home offices can only do SMBv1 but SMBv2 is minimum, so FTP is fallback), but as @fabianfrz dislikes this, I do, too. Preferable, it shall be WebDAVS for direct secure sync.

Unencrypted communication of course should never be allowed and if certificate issues are the topic, installing self-signed certificates under Trust should solve this (but as the Trust certificates are not installed in the system, one should be able to select the certificate similar to HAproxy for validation.) Of course, valid certificates like Lets-Encrypt should be allowed without these steps.

Another good idea of @fabianfrz was to use SCP/SSH but he stated that this is too mighty for base installation, but it could be a plugin. Of course, SCP/SSH might not have many real use cases as far as I would think...

**Comment by fabianfrz** [comment]
@Addy90 you misunderstood - my implementation of SFTP backup needs libs, I won't get for deployment in core (keep the core as small as possible).

Here is a work in progress PR for Nextcloud: #2251

**Comment by Addy90** [comment]
Ah right, my fault, it was the SFTP variant that needed additional libraries, not the SCP variant which is already in core as command...
But to be true, in the end, there is not much difference for a single backup file between SCP and SFTP, right? I mean, the advantages of SFTP are not needed for a single backup file... so maybe it could be an option... but nevertheless, the WebDAVS variant should have more impact than SCP...

I really like your work in progress. Can this be tested via opnsense-patch?

**Comment by fabianfrz** [comment]
yes it can - you need to apply the commits in order and need the following params:
`-a fabianfrz -c core`. Please note that the last commit only contains the chmod so if this cannot be applied, you need to do a chmod by yourself.

**Comment by fabianfrz** [comment]
Screenshot: 
![screenshot_20180309_182135](https://user-images.githubusercontent.com/3713914/37220463-0aca55ce-23c7-11e8-9d8e-3d39ef51c389.png)
User: normal username - I created a service account only for that for testing on my production server
PW: Create an app password and use it

**Comment by fabianfrz** [comment]
Regarding SCP:
Here is some code: https://github.com/opnsense/plugins/tree/master/sysutils/scp-backup

**Comment by Addy90** [comment]
Thanks, I am currently trying, too! patch worked!
Problem in line 320 of diag_backup.php for me, the regex does not permit directories with capital letters and no subdirectories (I have a directory /Backups/opnsense/ for my backup here, as I have other backups like smartphones, etc.)
I would not have found out without the code as the error just is "Nextcloud communication failure", maybe the regex is too restrictive?

**Comment by fabianfrz** [comment]
'i' flag should help

**Comment by fabianfrz** [comment]
The error message means that the upload script failed. You can call it manually via the shell to get the error (I added code to write the error to stdout)

**Comment by fabianfrz** [comment]
latest patch allows capital letters

**Comment by Addy90** [comment]
hmm yes, but no subdirectories... which then has a problem with the concatenation of the directory in the remote.php URL (I don't guess ../ is permitted in Nextcloud, is it?).
Somehow I still have a connection error problem, still investigating where the problem lies...

**Comment by Addy90** [comment]
Ah, problem was I had to restart configd first... my fault, did not read [the docs](https://docs.opnsense.org/development/backend/configd.html) careful enough.
Backup worked :D not in the folder I wanted, yet, but it worked :D

**Comment by fabianfrz** [comment]
please note that you can create a cron task if you like.

**Comment by Addy90** [comment]
This is fantastic! Still needs some polishing, @AdSchellevis is suggesting on the pull request already some, but this is fantastic. I will start using this on all three locations I have OPNsense running currently as soon as this is "production" ready.

**Comment by Addy90** [comment]
Oh, I think one bug is missing: The backup file has : (double colon) in it, which is a forbidden sign on windows, this is why the desktop client does not sync the backup file. Should be fixed, too...

**Comment by fabianfrz** [comment]
Thats not a bug on my side as a ":" is perfectly fine on Linux and BSD. @fichtner @AdSchellevis  Do you have a suggestion for another format that does not break Windows compatibility and is still readable?

**Comment by fichtner** [comment]
what is ":" for? the config backups don't use this:

```
# ls -lah /conf/backup
drwxr-xr-x  2 root  wheel   5.0K Mar  9 15:01 .
drwxr-xr-x  4 root  wheel   512B Mar  9 18:00 ..
-rw-r--r--  1 root  wheel   207K Feb 14 20:59 config-1518641974.3673.xml
-rw-r--r--  1 root  wheel   207K Feb 14 21:19 config-1518643147.8936.xml
-rw-r--r--  1 root  wheel   206K Feb 15 16:20 config-1518711648.9022.xml
-rw-r--r--  1 root  wheel   206K Feb 15 16:20 config-1518711650.7505.xml
-rw-r--r--  1 root  wheel   206K Feb 15 16:21 config-1518711698.7845.xml
```

**Comment by fabianfrz** [comment]
my uploads contain the upload time in the common format: Hour:Minute:Second

**Comment by fabianfrz** [comment]
@fichtner 
![screenshot_20180309_175829](https://user-images.githubusercontent.com/3713914/37225068-bce8eaaa-23d5-11e8-9210-88130953690e.png)

**Comment by brandonmreeves** [comment]
When will the scp-backup plugin be generally available via the opnsense gui? Ive been watching and would like to use it.

Thanks

**Comment by fabianfrz** [comment]
no, at some point it will be moved to core

**Comment by fabianfrz** [comment]
#2289 is close to a release
